File: station_service/api/schemas/manual.py
"""
Manual Control API Schema Definitions.

Defines request/response models for manual hardware control and
command introspection endpoints.
All responses use camelCase field names in JSON output via APIBaseModel.
"""

from typing import Any, Dict, List, Optional

from pydantic import Field

from station_service.api.schemas.base import APIBaseModel


class ParameterInfo(APIBaseModel):
    """Parameter definition for a hardware command."""

    name: str = Field(..., description="Parameter name")
    display_name: str = Field(..., description="Display name for UI")
    type: str = Field(..., description="Parameter type: string, number, boolean, select")
    required: bool = Field(False, description="Whether parameter is required")
    default: Optional[Any] = Field(None, description="Default value if not required")
    unit: Optional[str] = Field(None, description="Unit of measurement")
    min: Optional[float] = Field(None, description="Minimum value for numbers")
    max: Optional[float] = Field(None, description="Maximum value for numbers")
    options: Optional[List[Dict[str, Any]]] = Field(
        None, description="Options for select type"
    )
    description: Optional[str] = Field(None, description="Parameter description")


class CommandInfo(APIBaseModel):
    """Hardware command information."""

    name: str = Field(..., description="Command/method name")
    display_name: str = Field(..., description="Display name for UI")
    description: str = Field("", description="Command description")
    category: str = Field(
        "diagnostic",
        description="Category: measurement, control, configuration, diagnostic",
    )
    parameters: List[ParameterInfo] = Field(
        default_factory=list, description="Command parameters"
    )
    return_type: str = Field("Any", description="Return type")
    return_unit: Optional[str] = Field(None, description="Unit of return value")
    is_async: bool = Field(True, serialization_alias="async", description="Whether command is async")


class HardwareCommandsResponse(APIBaseModel):
    """Response containing available commands for a hardware device."""

    hardware_id: str = Field(..., description="Hardware device ID")
    driver: str = Field(..., description="Driver class name")
    connected: bool = Field(..., description="Whether device is connected")
    commands: List[CommandInfo] = Field(..., description="Available commands")


class HardwareDetailedStatus(APIBaseModel):
    """Detailed hardware status information."""

    id: str = Field(..., description="Hardware device ID")
    driver: str = Field(..., description="Driver class name")
    status: str = Field(..., description="Connection status")
    connected: bool = Field(..., description="Whether connected")
    config: Dict[str, Any] = Field(default_factory=dict, description="Configuration")
    info: Dict[str, Any] = Field(default_factory=dict, description="Device info")
    last_error: Optional[str] = Field(None, description="Last error message")


class ManualStepConfig(APIBaseModel):
    """Manual mode configuration for a sequence step."""

    skippable: bool = Field(True, description="Can be skipped in manual mode")
    auto_only: bool = Field(False, description="Only runs in automatic mode")
    prompt: Optional[str] = Field(None, description="Confirmation prompt")
    pause_before: bool = Field(False, description="Pause before execution")
    pause_after: bool = Field(False, description="Pause after execution")
    parameter_overrides: List[str] = Field(
        default_factory=list, description="Parameters that can be overridden"
    )


class ManualStepInfo(APIBaseModel):
    """Step information for manual sequence execution."""

    name: str = Field(..., description="Step name")
    display_name: str = Field(..., description="Display name")
    order: int = Field(..., description="Execution order")
    timeout: float = Field(60.0, description="Timeout in seconds")
    manual: ManualStepConfig = Field(
        default_factory=ManualStepConfig, description="Manual mode config"
    )
    status: str = Field("pending", description="Current status")
    result: Optional[Dict[str, Any]] = Field(None, description="Step result")
    duration: Optional[float] = Field(None, description="Execution duration")


class ManualStepRequest(APIBaseModel):
    """Request body for manual step execution."""

    parameters: Optional[Dict[str, Any]] = Field(
        None, description="Parameter overrides for the step"
    )


class CommandPreset(APIBaseModel):
    """Saved command preset for quick access."""

    id: str = Field(..., description="Preset ID")
    name: str = Field(..., description="Preset name")
    hardware_id: str = Field(..., description="Hardware device ID")
    command: str = Field(..., description="Command name")
    params: Dict[str, Any] = Field(default_factory=dict, description="Command parameters")
    created_at: str = Field(..., description="Creation timestamp")


class CommandPresetCreate(APIBaseModel):
    """Request body for creating a command preset."""

    name: str = Field(..., description="Preset name")
    hardware_id: str = Field(..., description="Hardware device ID")
    command: str = Field(..., description="Command name")
    params: Dict[str, Any] = Field(default_factory=dict, description="Command parameters")

File: station_ui/src/types/manual.ts
/**
 * Manual Control Type Definitions.
 *
 * Types for enhanced manual control page functionality.
 */

/**
 * Parameter definition for a hardware command.
 */
export interface CommandParameter {
  /** Parameter name */
  name: string;
  /** Display name for UI */
  displayName: string;
  /** Parameter type */
  type: 'string' | 'number' | 'boolean' | 'select' | 'range';
  /** Whether parameter is required */
  required: boolean;
  /** Default value */
  default?: unknown;
  /** Unit of measurement */
  unit?: string;
  /** Minimum value for numbers */
  min?: number;
  /** Maximum value for numbers */
  max?: number;
  /** Options for select type */
  options?: Array<{ value: unknown; label: string }>;
  /** Parameter description */
  description?: string;
}

/**
 * Hardware command definition.
 */
export interface CommandInfo {
  /** Command/method name */
  name: string;
  /** Display name for UI */
  displayName: string;
  /** Command description */
  description: string;
  /** Category: measurement, control, configuration, diagnostic */
  category: 'measurement' | 'control' | 'configuration' | 'diagnostic';
  /** Command parameters */
  parameters: CommandParameter[];
  /** Return type */
  returnType: string;
  /** Return unit */
  returnUnit?: string;
  /** Whether command is async */
  async: boolean;
}

/**
 * Response containing available commands for a hardware device.
 */
export interface HardwareCommandsResponse {
  /** Hardware device ID */
  hardwareId: string;
  /** Driver class name */
  driver: string;
  /** Whether device is connected */
  connected: boolean;
  /** Available commands */
  commands: CommandInfo[];
}

/**
 * Detailed hardware status.
 */
export interface HardwareDetailedStatus {
  /** Hardware device ID */
  id: string;
  /** Driver class name */
  driver: string;
  /** Connection status */
  status: 'connected' | 'disconnected' | 'error';
  /** Whether connected */
  connected: boolean;
  /** Configuration */
  config: Record<string, unknown>;
  /** Device info */
  info: Record<string, unknown>;
  /** Last error message */
  lastError?: string;
}

/**
 * Manual step configuration.
 */
export interface ManualStepConfig {
  /** Can be skipped in manual mode */
  skippable: boolean;
  /** Only runs in automatic mode */
  autoOnly: boolean;
  /** Confirmation prompt */
  prompt?: string;
  /** Pause before execution */
  pauseBefore: boolean;
  /** Pause after execution */
  pauseAfter: boolean;
  /** Parameters that can be overridden */
  parameterOverrides: string[];
}

/**
 * Step information for manual execution.
 */
export interface ManualStepInfo {
  /** Step name */
  name: string;
  /** Display name */
  displayName: string;
  /** Execution order */
  order: number;
  /** Timeout in seconds */
  timeout: number;
  /** Manual mode config */
  manual: ManualStepConfig;
  /** Current status */
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped';
  /** Step result */
  result?: Record<string, unknown>;
  /** Execution duration */
  duration?: number;
}

/**
 * Command preset for quick access.
 */
export interface CommandPreset {
  /** Preset ID */
  id: string;
  /** Preset name */
  name: string;
  /** Hardware device ID */
  hardwareId: string;
  /** Command name */
  command: string;
  /** Command parameters */
  params: Record<string, unknown>;
  /** Creation timestamp */
  createdAt: string;
}

/**
 * Request body for creating a command preset.
 */
export interface CreatePresetRequest {
  /** Preset name */
  name: string;
  /** Hardware device ID */
  hardwareId: string;
  /** Command name */
  command: string;
  /** Command parameters */
  params: Record<string, unknown>;
}

/**
 * Result history entry.
 */
export interface ResultHistoryEntry {
  /** Unique ID */
  id: string;
  /** Hardware device ID */
  hardware: string;
  /** Command executed */
  command: string;
  /** Command parameters */
  params: Record<string, unknown>;
  /** Execution result */
  result: unknown;
  /** Whether execution succeeded */
  success: boolean;
  /** Execution duration in ms */
  duration: number;
  /** Timestamp */
  timestamp: Date;
  /** Unit for numeric results */
  unit?: string;
}

/**
 * Manual control state for Zustand store.
 */
export interface ManualControlState {
  /** Selected batch ID */
  selectedBatchId: string | null;
  /** Selected hardware ID */
  selectedHardwareId: string | null;
  /** Selected command */
  selectedCommand: CommandInfo | null;
  /** Parameter values */
  parameterValues: Record<string, unknown>;
  /** Result history */
  resultHistory: ResultHistoryEntry[];
  /** Command presets */
  presets: CommandPreset[];

  // Manual sequence mode
  /** Whether manual sequence mode is enabled */
  manualSequenceMode: boolean;
  /** Sequence steps */
  sequenceSteps: ManualStepInfo[];
  /** Current step index */
  currentStepIndex: number;
  /** Parameter overrides per step */
  stepOverrides: Record<string, Record<string, unknown>>;

  // Actions
  selectDevice: (batchId: string | null, hardwareId: string | null) => void;
  selectCommand: (command: CommandInfo | null) => void;
  setParameterValue: (name: string, value: unknown) => void;
  setParameterValues: (values: Record<string, unknown>) => void;
  addResultToHistory: (entry: Omit<ResultHistoryEntry, 'id' | 'timestamp'>) => void;
  clearHistory: () => void;
  addPreset: (preset: CommandPreset) => void;
  removePreset: (presetId: string) => void;

  // Manual sequence actions
  setManualSequenceMode: (enabled: boolean) => void;
  setSequenceSteps: (steps: ManualStepInfo[]) => void;
  updateStepStatus: (
    stepName: string,
    status: ManualStepInfo['status'],
    result?: Record<string, unknown>,
    duration?: number
  ) => void;
  setCurrentStepIndex: (index: number) => void;
  setStepOverride: (stepName: string, overrides: Record<string, unknown>) => void;
  resetSequence: () => void;
}
